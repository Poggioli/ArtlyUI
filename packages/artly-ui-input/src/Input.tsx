import { forwardRef } from "@artly-ui/core";
import { ChangeEvent, FocusEvent, useMemo, useState } from "react";
import {
  StyledFieldSet,
  StyledFlexContainer,
  StyledFlexContainerInput,
  StyledInput,
  StyledInputLabel,
  StyledLegend,
  StyledLegendSpan,
} from "./styles";
import { InputProps } from "./types";

const Input = forwardRef<"input", InputProps>(
  (
    {
      label,
      id,
      css,
      onFocus,
      onBlur,
      onChange,
      shape,
      placeholder,
      required,
      color,
      ...props
    },
    forwardedRef
  ) => {
    const newId = useMemo(() => id ?? Date.now(), [id]);
    const newLabel = useMemo(
      () => (!!required ? `${label} *` : label),
      [label, required]
    );
    const [hasFocus, setHasFocus] = useState(!!props.autoFocus);
    const [hasValue, setHasValue] = useState(!!props.value);
    const placeholderValue = useMemo(
      () => (hasFocus && !hasValue ? placeholder : ""),
      [hasFocus, hasValue, placeholder]
    );

    const handleOnFocus = (event: FocusEvent<HTMLInputElement, Element>) => {
      setHasFocus((previousState) => !previousState);
      if (onFocus) {
        onFocus(event);
      }
    };

    const handleOnBlur = (event: FocusEvent<HTMLInputElement, Element>) => {
      setHasFocus((previousState) => !previousState);
      if (onBlur) {
        onBlur(event);
      }
    };

    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
      setHasValue(!!event.target.value);
      if (onChange) {
        onChange(event);
      }
    };

    return (
      <StyledFlexContainer direction="column" css={css} color={color}>
        <StyledInputLabel
          htmlFor={newId}
          hasValue={hasValue}
          hasFocus={hasFocus}
          shape={shape}
        >
          {newLabel}
        </StyledInputLabel>
        <StyledFlexContainerInput align="center">
          <StyledInput
            {...props}
            ref={forwardedRef}
            id={newId}
            shape={shape}
            onFocus={handleOnFocus}
            onBlur={handleOnBlur}
            onChange={handleChange}
            placeholder={placeholderValue}
            color={color}
          />
          <StyledFieldSet shape={shape} aria-hidden>
            <StyledLegend hasValue={hasValue || hasFocus}>
              <StyledLegendSpan variant="auxiliary">
                {newLabel}
              </StyledLegendSpan>
            </StyledLegend>
          </StyledFieldSet>
        </StyledFlexContainerInput>
      </StyledFlexContainer>
    );
  }
);

export { Input };
