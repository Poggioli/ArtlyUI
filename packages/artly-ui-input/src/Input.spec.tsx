import { render } from "@testing-library/react";
import { Input } from "./Input";

describe("Core component / Input", () => {
  it(`GIVEN a Input component
    WHEN render component
    SHOULD occurs render normaly`, () => {
    const rendered = render(<Input />);
    expect(rendered.container).toBeTruthy();
  });
});
