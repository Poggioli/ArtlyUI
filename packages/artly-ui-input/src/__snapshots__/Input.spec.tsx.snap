// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core component / Input GIVEN a Input compoenent
      WHEN has autofocus
      THEN should render normaly 1`] = `
<div>
  <div
    class="artly-UI-c-bXbWpx artly-UI-c-hIFJPS artly-UI-c-hIFJPS-haMmXz-color-unstyled"
    direction="column"
  >
    <label
      class="artly-UI-c-chWhuJ artly-UI-c-chWhuJ-jprvTt-variant-body1 artly-UI-c-chWhuJ-cmVlgk-align-left artly-UI-c-chWhuJ-iiBiFgP-css artly-UI-c-hjMDXt artly-UI-c-hjMDXt-dikOXF-shape-standard"
      for="12345"
    >
      Enter your email
    </label>
    <div
      align="center"
      class="artly-UI-c-bXbWpx artly-UI-c-bSMcqW"
    >
      <input
        class="artly-UI-c-gJkmaK artly-UI-c-gJkmaK-bzjyrz-shape-standard artly-UI-c-gJkmaK-dKQVtd-color-unstyled"
        id="12345"
        placeholder=""
      />
      <fieldset
        aria-hidden="true"
        class="artly-UI-c-jaKNMU artly-UI-c-jaKNMU-bbgQrM-shape-standard"
      >
        <legend
          class="artly-UI-c-cKPsdG"
        >
          <span
            class="artly-UI-c-chWhuJ artly-UI-c-jijTon artly-UI-c-chWhuJ-bzGfXE-variant-auxiliary artly-UI-c-chWhuJ-cmVlgk-align-left"
          >
            Enter your email
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`Core component / Input GIVEN a Input compoenent
      WHEN has id
      THEN should render normaly 1`] = `
<div>
  <div
    class="artly-UI-c-bXbWpx artly-UI-c-hIFJPS artly-UI-c-hIFJPS-haMmXz-color-unstyled"
    direction="column"
  >
    <label
      class="artly-UI-c-chWhuJ artly-UI-c-chWhuJ-jprvTt-variant-body1 artly-UI-c-chWhuJ-cmVlgk-align-left artly-UI-c-chWhuJ-iiBiFgP-css artly-UI-c-hjMDXt artly-UI-c-hjMDXt-dikOXF-shape-standard"
      for="12345"
    >
      Enter your email
    </label>
    <div
      align="center"
      class="artly-UI-c-bXbWpx artly-UI-c-bSMcqW"
    >
      <input
        class="artly-UI-c-gJkmaK artly-UI-c-gJkmaK-bzjyrz-shape-standard artly-UI-c-gJkmaK-dKQVtd-color-unstyled"
        id="12345"
        placeholder=""
      />
      <fieldset
        aria-hidden="true"
        class="artly-UI-c-jaKNMU artly-UI-c-jaKNMU-bbgQrM-shape-standard"
      >
        <legend
          class="artly-UI-c-cKPsdG"
        >
          <span
            class="artly-UI-c-chWhuJ artly-UI-c-jijTon artly-UI-c-chWhuJ-bzGfXE-variant-auxiliary artly-UI-c-chWhuJ-cmVlgk-align-left"
          >
            Enter your email
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`Core component / Input GIVEN a Input compoenent
      WHEN has no value
      AND has focus
      THEN should render placeholder 1`] = `
<div>
  <div
    class="artly-UI-c-bXbWpx artly-UI-c-hIFJPS artly-UI-c-hIFJPS-haMmXz-color-unstyled"
    direction="column"
  >
    <label
      class="artly-UI-c-chWhuJ artly-UI-c-chWhuJ-jprvTt-variant-body1 artly-UI-c-chWhuJ-cmVlgk-align-left artly-UI-c-chWhuJ-iiBiFgP-css artly-UI-c-hjMDXt artly-UI-c-hjMDXt-dikOXF-shape-standard"
      for="12345"
    >
      Enter your email
    </label>
    <div
      align="center"
      class="artly-UI-c-bXbWpx artly-UI-c-bSMcqW"
    >
      <input
        class="artly-UI-c-gJkmaK artly-UI-c-gJkmaK-bzjyrz-shape-standard artly-UI-c-gJkmaK-dKQVtd-color-unstyled"
        id="12345"
        placeholder=""
      />
      <fieldset
        aria-hidden="true"
        class="artly-UI-c-jaKNMU artly-UI-c-jaKNMU-bbgQrM-shape-standard"
      >
        <legend
          class="artly-UI-c-cKPsdG"
        >
          <span
            class="artly-UI-c-chWhuJ artly-UI-c-jijTon artly-UI-c-chWhuJ-bzGfXE-variant-auxiliary artly-UI-c-chWhuJ-cmVlgk-align-left"
          >
            Enter your email
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`Core component / Input GIVEN a Input compoenent
      WHEN has value
      THEN should render normaly 1`] = `
<div>
  <div
    class="artly-UI-c-bXbWpx artly-UI-c-hIFJPS artly-UI-c-hIFJPS-haMmXz-color-unstyled"
    direction="column"
  >
    <label
      class="artly-UI-c-chWhuJ artly-UI-c-chWhuJ-jprvTt-variant-body1 artly-UI-c-chWhuJ-cmVlgk-align-left artly-UI-c-chWhuJ-iiBiFgP-css artly-UI-c-hjMDXt artly-UI-c-hjMDXt-dikOXF-shape-standard artly-UI-c-hjMDXt-eqqjGp-cv"
      for="12345"
    >
      Enter your email
    </label>
    <div
      align="center"
      class="artly-UI-c-bXbWpx artly-UI-c-bSMcqW"
    >
      <input
        class="artly-UI-c-gJkmaK artly-UI-c-gJkmaK-bzjyrz-shape-standard artly-UI-c-gJkmaK-dKQVtd-color-unstyled"
        id="12345"
        placeholder=""
        value="email@email.com"
      />
      <fieldset
        aria-hidden="true"
        class="artly-UI-c-jaKNMU artly-UI-c-jaKNMU-bbgQrM-shape-standard"
      >
        <legend
          class="artly-UI-c-cKPsdG artly-UI-c-cKPsdG-boYBol-hasValue-true"
        >
          <span
            class="artly-UI-c-chWhuJ artly-UI-c-jijTon artly-UI-c-chWhuJ-bzGfXE-variant-auxiliary artly-UI-c-chWhuJ-cmVlgk-align-left"
          >
            Enter your email
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
